### ISSUE TITLE
Theia framework fails to supply Help with entries for the IDE's commands

### ISSUE THREAD

**Issue by chrisjj** [eclipse-theia/theia#7160]
(Since the change of title has obfuscated the meaning of this post, I'm restoring the original.)

Hence that the Theia example IDE's Help has virtually no info on the commands.

**Update** (@vince-fugnitto)

The goal is to provide users with the ability to view all available registered commands present in the framework, alongside helpful descriptions of their functionality. Such a reference can then be easily added as a menu entry under the `Help` menu for users to reference.

**Comment by vince-fugnitto** [comment]
@chrisjj are you referring to a menu entry such as 'Help' > 'Keyboard Shortcuts Reference' in VS Code which links to the following [document](https://code.visualstudio.com/shortcuts/keyboard-shortcuts-macos.pdf)?

**Comment by chrisjj** [comment]
No. As documentation for an IDE's commands, that example is woefully incomplete.

Help > Reference declaring and describing the commands would do fine. Without something like that, there's no way the user can find out what commands are available and what they do.

**Comment by vince-fugnitto** [comment]
> Help > Reference declaring and describing the commands would do fine. Without something like that, there's no way the user can find out what commands are available and what they do.

Thank you for clarifying. I believe it would be helpful to support `descriptions` for registered commands, so that users can not only get a more in depth description of an actual command, but also use such a parameter to generate documentation in an automated way. 

At the moment, we (including vscode), try to be as clear as possible when naming commands so they are easy to understand (ex: `search: clear search results`). However, I can also understand a user wishing to view a centralized place for all commands with a brief description of what the command actually does and in which context. 

As always, pull-requests are welcome!

**Comment by chrisjj** [comment]
> I believe it would be helpful to support `descriptions` for registered commands, so that users can not only get a more in depth description of an actual command, but also use such a parameter to generate documentation.

Indeed.

Plus there would need to be project-wide policy that makes these descriptions mandatory, preventing Help being eroded by addition of commands lacking descriptions.

I suggest this should be done before V1.0.0 release. It might be harder to gain project-wide acceptance later.

**Comment by vince-fugnitto** [comment]
> > I believe it would be helpful to support `descriptions` for registered commands, so that users can not only get a more in depth description of an actual command, but also use such a parameter to generate documentation.
> 
> I do too.
> 
> Plus there would need to be project-wide policy that makes these descriptions mandatory, to prevent the initially complete Help being undermined by addition of commands lacking descriptions.

A policy would not be necessary, making the description mandatory in the code would be enough. This is of course a breaking change that I don’t believe would be immediately worth the effort (it would force changes to downstream projects). If one cannot understand what a command does based on the label (in the command palette), they can explore what the command performs through experimentation. This is the approach I adopted in vscode and did not find any problems.

> I suggest this should all be done before V1.0.0 release. It may well be much harder to gain project-wide acceptance later.

We already discussed the necessary items for the 1.0.0 release by all main maintainers and companies, and while it’d be nice to have documentation of available methods it is definitely not a priority (especially since commands are mostly self-explanatory). A product can always document their own commands if they so chose.

**Comment by chrisjj** [comment]
> A policy would not be necessary, making the description mandatory in the code would be enough.

If that policy could take the form of build control that made description mandatory in the code, great.

> This is of course a breaking change ...  (it would force changes to downstream projects).

Yes and yes, and hence that advantage of doing it before V1. The more sub-standard Theia apps there are, the greater the resistance to raising the standard. And this cascades - sub-standard projects attract more of the same and discourage developers with higher standards.

>  If one cannot understand what a command does based on the label (in the command palette), they can explore what the command performs through experimentation.

And then perhaps they can read the source?? :)

> A product can always document their own commands if they so chose.

Quite why anyone interested in proper documentation would make a distinction, I do not know. 

They'd end up documenting the core commands and then be chasing that in each major update going forward for ever.

**Comment by vince-fugnitto** [comment]
> Yes and yes, and hence that advantage of doing it before V1. The more sub-standard Theia apps there are, the greater the resistance to raising the standard. And this cascades - sub-standard projects attract more of the same and discourage developers with higher standards.

As mentioned this will not be part of v1.0.0. There are other priorities for the project, and there are other ways that commands can be documented today without the use of breaking changes. I doubt this will discourage developers with higher standards (one can always extend the base command to enforce a mandatory `description`)

> The test of that 'solution' is whether Theia would be happy to declare it in the place the user would look for actual IDE documentation. I'm guessing and hoping it would be too ashamed.

Documenting a list of available commands the IDE supports is an enhancement at this moment.
The framework has prioritized other features/documentation which are targeted to application developers (developers using the framework to build their own products). Application developers are free to document their own IDEs for the moment until a list is documented in the framework.

If one cannot infer what a command does by it's `label`, they are free to test out it's features (by executing the command), or by viewing it's sources. All these methods are valid workarounds for the moment and thus mean this is not a priority. It can be done someday, but it does not necessarily need to be done today.

**Comment by chrisjj** [comment]
> As mentioned this will not be part of v1.0.0. 

That wasn't mentioned.

> There are other priorities for the project

The fact there is already more than one is no grounds for not adding another.

> and there are other ways that commands can be documented today 

Evidently not good enough, since no-one is using them.

>without the use of breaking changes.

The only breaking change in requiring Help entries and providing them for core commands is to any custom commands added by IDEs, right?

In which case, that's fine, right? Since Theia is still in pre-V1 beta state,

Whereas procrastinating threatens breakage in the production release.

> (one can always extend the base command to enforce a mandatory `description`)

The problem with that is it requires the user to go back and add documentation for core commands written by other developers. Bad idea. Writing a command's documentation should lie with the command's implementer.

> If one cannot infer what a command does by it's `label`, they are free to test out it's features (by executing the command), or by viewing it's sources.

That answer could save a lot of wasted time by any prospective developer looking for Theia command documentation and landing here, so I'm highlighting it:

 "If one cannot infer what a command does by it's `label`, they are free to test out it's features (by executing the command), or by viewing it's sources. "

Thanks for the clarification.

> All these methods are valid workarounds for the moment and thus mean this is not a priority.

And there was me thinking V1 was imminent.

Thanks for the clarification.

**Comment by vince-fugnitto** [comment]
> > As mentioned this will not be part of v1.0.0.
> 
> That wasn't mentioned.
> 

As mentioned in this https://github.com/eclipse-theia/theia/issues/7160#issuecomment-586649445, "We already discussed the necessary items for the 1.0.0 release by all main maintainers and companies, and while it’d be nice to have documentation of available methods it is definitely not a priority."

The project, as I previously mentioned, has a list of priorities for v1. This means that nice-to-haves such as documenting commands in some form will likely not be part of it since v1 has a fixed date.

> > There are other priorities for the project
> 
> The fact there is already more than one is no grounds for not adding another.

We have a fixed date for the release of v1.0.0. We can try to fit nice-to-haves in for the release but there is no guarantee that those items will make it. We have our main priorities and those need to be addressed first.

> Whereas procrastinating threatens breakage in the production release.

No one is procrastinating. We have priorities like any project. If this is a priority for you or your application, feel free to open a pull-request.

**Comment by chrisjj** [comment]
> As mentioned in this [#7160 (comment)](https://github.com/eclipse-theia/theia/issues/7160#issuecomment-586649445)

I saw that, thanks.  It does not mention "this will not be part of v1.0.0.".

You've sine stated "this will not be part of v1.0.0." and thanks for being open on that.

**Comment by marcdumais-work** [comment]
@chrisjj We, the project members, reserve the right to reword, prioritise and generally manage the issues in this repository, as we see fit. 

I believe the title of this issue here is subjective and unnecessarily negative - any feature that Theia does not immediately have could be seen as a "failure" of some sort. @vince-fugnitto attempted to improve upon the original title but you reverted the change. We have little time to play these games.

Closing as duplicate of #7203
